### PR TITLE
Lazy latest dt

### PR DESCRIPTION
**Pull request by ehebert**
ENH: Provide method for last trade dt.

    Add a data_portal method `get_last_traded_dt` which the dt passed to the
    method if there is a trade on that bar, or the previous dt with a trade
    if there is no trade on the current dt.

    This method will be used to proved a `.dt` lookup on SidView, but the
    connection to the SidView is not made with this patch.

    To implement the get_last_traded_dt a corresponding method
    was added to the minute bar reader, with a supporting DatetimeIndex
    which matches the data sets indices. Also, change
    `_find_position_of_minute` to use the index, since the lookup is faster
    than calculating a Timestamp and doing the offset calculation every
    time. (As seen by the improvement in the ffill tests in test_dataportal.)

    Also, change history tests to use a shared data_portal to avoid setup
    cost of repeatedly creating the lookup table in the reader.

### CODE REVIEW

File: zipline/gens/sim_engine.pyx [hunk 1]
```diff
@@ -86,7 +83,7 @@ cdef class MinuteSimulationClock:
                     yield minute, MINUTE_END
 
             if not minute_emission:
-                yield day, DAY_END
+                yield minutes[-1], DAY_END
```
Comment (jbredeche) [inline zipline/gens/sim_engine.pyx]: why this change?  daily mode needs days, right?  (ie, 00:00, not 16:00)

Comment (ehebert) [inline zipline/gens/sim_engine.pyx]: This is  the minute mode clock. In both daily and minute emissions, the prices used for the day end should be at the market_close.

Noticed this when using `get_loc` instead of a combination of searchsorted for getting the minute position. From reading the code it may have been that the first minute of the day was being used for the last sale prices for risk calculations, but have not confirmed.

File: zipline/data/data_portal.py [hunk 2]
```diff
@@ -234,6 +212,16 @@ def _get_ctable(self, asset):
...
+    def get_last_traded_dt(self, asset, dt, data_frequency):
+        """
+        Given an asset and dt, returns the last traded dt from the viewpoint
+        of the given dt.
+
+        If there is a trade on the dt, the answer is dt provided.
+        """
+        if data_frequency == 'minute':
+            return self._equity_minute_reader.get_last_traded_dt(asset, dt)
+
```
Comment (jbredeche) [inline zipline/data/data_portal.py]: should we blow up on other data frequencies (for now)?

Comment (ehebert) [inline zipline/data/data_portal.py]: Looking at this again, with wiring this to SidView, we'll have to implement for daily.
